acoustid/future/fpindex/updater.py

```diff
@@ -0,0 +1,179 @@
+    return [x.strip() for x in value.strip("{}").split(",")]
+
+
+async def create_replication_slot(conn: asyncpg.Connection) -> tuple[bool, str | None]:
+    slot_exists = await conn.fetchval(
+        """
+        SELECT COUNT(*) > 0
+        FROM pg_replication_slots
+        WHERE slot_name = $1
+        """,
+        SLOT_NAME,
+    )
+
+    if slot_exists:
+        logger.info("Replication slot %s already exists", SLOT_NAME)
+        return False, None
+
+    logger.info("Creating snapshot")
+    snapshot_id = await conn.fetchval("SELECT pg_export_snapshot()")
+
+    logger.info("Creating replication slot %s", SLOT_NAME)
+    await conn.execute(
+        """
+            SELECT * FROM pg_create_logical_replication_slot($1, 'wal2json', temporary => true)
+        """,
+        SLOT_NAME,
+    )
+    return True, snapshot_id
+
+
+async def drop_replication_slot(conn: asyncpg.Connection) -> None:
+    logger.info("Dropping replication slot %s", SLOT_NAME)
+    await conn.execute(
+        "SELECT * FROM pg_drop_replication_slot($1)",
+        SLOT_NAME,
+    )
+
+
+async def load_initial_data(conn: asyncpg.Connection) -> None:
+    """Load initial fingerprint data using a consistent snapshot"""
+
+    async with AsyncExitStack() as exit_stack:
+        await exit_stack.enter_async_context(
+            conn.transaction(isolation="repeatable_read", readonly=True)
+        )
+
+        needs_import, _ = await create_replication_slot(conn)
+        if not needs_import:
+            return
+
+        async def drop_replication_slot_on_error(*exc_details) -> None:
+            if exc_details[0] is not None:
+                await drop_replication_slot(conn)
+
+        exit_stack.push_async_exit(drop_replication_slot_on_error)
+
+        total_count = await conn.fetchval(
+            """
+            SELECT reltuples::bigint AS estimate
+            FROM pg_class
+            WHERE relname = 'fingerprint'
+            """
+        )
+        logger.info("Found approximately %s fingerprints", total_count)
+
+        cursor = conn.cursor(
+            """
+            SELECT id, fingerprint::text
+            FROM fingerprint
+            """
+        )
+        loaded_count = 0
+        async for row in cursor:
+            await asyncio.sleep(0.1)
+            logger.info(
+                "Processing fingerprint %d - %s",
+                row["id"],
+                decode_postgres_array(row["fingerprint"]),
+            )
+            loaded_count += 1
+            if loaded_count % 10 == 0:
+                logger.info(
+                    "Processed %d/%d fingerprints (%.2f%%)",
+                    loaded_count,
+                    total_count,
+                    (loaded_count / total_count * 100),
+                )
+
+    logger.info(f"Initial data load complete. Processed {loaded_count} fingerprints.")
+
+
+async def replicate_from_pg() -> None:
+    async with AsyncExitStack() as exit_stack:
```

Comment:
@coderabbitai please explain how there could be a gap between initial load and replication? The load is done in a snapahot transaction, and the slot is created from inside the same transaction.